### ISSUE TITLE
Issues with Commissions + Simple Shipping + PayPal Adaptive Payments

### ISSUE THREAD

**Issue by pippinsplugins** [easydigitaldownloads/EDD-Commissions#125]
There are several issues with the integration between Simple Shipping, Commissions, and PayPal Adaptive Payments.
1. The include/exclude shipping fee options do not work when paying out commissions instantly through PayPal Adaptive Payments.
2. When the commission rate is set to 100%, the redirect to PayPal fails with an `amount must be greater than zero` error.

**Comment by pippinsplugins** [comment]
This can still fail when calculation base is subtotal.

**Comment by pippinsplugins** [comment]
@mintplugins Can you test this again?

**Comment by mintplugins** [comment]
**"Shipping Fees" setting causes `greater than zero` error:**

Under "Downloads" > "Settings" > "Comissions", if "Shipping Fees" are set to "Shipping fee not paid to recipient", everything works correctly. However, if set to any other option, it gives the `greater than zero ` error.

You'll get the `amount must be greater than zero` error if "Shipping Fees" are set to:
- "Shipping fee paid to recipient"
- "Ignore Shipping Fees"

**To Replicate:**

1. Set up a download with shipping and 2 commissions with 50%
<img width="296" alt="screen shot 2016-03-11 at 12 34 33 pm" src="https://cloud.githubusercontent.com/assets/1638128/13710140/a4bf7be4-e785-11e5-8938-5d498a20e328.png">

2. Under "Downloads" > "Settings" > "Commissions", set "Shipping Fees" to anything **other** than "Shipping fee not paid to recipient".

3. Set Paypal Adaptive Receivers as **storeowner@store.com|100**. Using Chained or Parallel does not matter for this.

4. Attempt a purchase.

**Other Info**
Installed Plugin Versions:

- EDD: 2.5.9 
- Commissions: 3.2.10 
- Paypal Adaptive Payments: 1.3
- Simple Shipping: 2.2.0

**Comment by mintplugins** [comment]
@pippinsplugins This ended up being something that I realized was an issue with Commissions and Shipping in general. It lead to me re-writing the shipping portion of the main commission functions. There are a lot of changes. My changes make the following work:

- Shipping wasn't calculated correctly for commissions if set to anything for "Reciever split" settings.
- Shipping with variable pricing options were not addressed or calculated for commissions.

These fixes also lead to the following enhancements:
- Each download can have a unique setting for commissions and shipping fees.
- Better definitions of "Receiver Split" settings: https://github.com/easydigitaldownloads/EDD-Commissions/issues/127

The main issue with Paypal Adaptive still needs to be addressed and I plan to get to it tomorrow. 

My main question now is whether I should open an issue for each of the above and separate my commits out to different branches or if I can commit them all as a single pull request for this issue. Which would be the correct action for this here? For me, ideally I'd like this issue to grow to contain all of the above as this issue lead me to doing all of that stuff.

**Comment by mintplugins** [comment]
Sidenote for discussion: I also believe that no shipping functions should actually be in the Commissions extension and all should be moved to Simple Shipping. I may need to add a couple of filter hooks to commissions to make that possible but to me that makes the most sense.

**Comment by pippinsplugins** [comment]
@mintplugins let's do it all as one issue since they're so closely related.

**Comment by mintplugins** [comment]
@pippinsplugins Here's the PR: https://github.com/easydigitaldownloads/EDD-Commissions/pull/130/files

**Comment by mintplugins** [comment]
What should be tested:

No Shipping:
- A single product with commissions
- A single product with variable prices and commissions

With Shipping:
- A single product with commissions and shipping
- A single product with variable prices and commissions

Also should do test purchases with a combination of products with Commissions, Commissions and Shipping, and no commissions.

**Comment by pippinsplugins** [comment]
@mintplugins I'm going to test this tomorrow.

**Comment by mintplugins** [comment]
@pippinsplugins 
You can hold off on testing for a bit - I have the customer doing some tests and there's a couple issues I still have to work out with it. I'll be adding more commits later today (hopefully)

**Comment by pippinsplugins** [comment]
:+1:

**Comment by mintplugins** [comment]
This issue will require a fix to this bug:
https://github.com/easydigitaldownloads/easy-digital-downloads/issues/4477

**Comment by pippinsplugins** [comment]
@mintplugins ready for more testing?

**Comment by mintplugins** [comment]
@pippinsplugins  Getting close. I have price_id's being attached to fees now here:
https://github.com/easydigitaldownloads/easy-digital-downloads/issues/4477

I just need to make it so that "edd_get_payment_meta_cart_details()" only attached the right fee to the right price ID now.

**Comment by pippinsplugins** [comment]
:+1: let me know if you need any direction

**Comment by mintplugins** [comment]
The issue that needs to be fixed with fees is ready to test here: https://github.com/easydigitaldownloads/easy-digital-downloads/pull/4494

**Comment by mintplugins** [comment]
For when the adaptive function gets re-written, make sure that the site_owner is taken out of the array passed to Paypal Adaptive if the mode is "Chained".

![img_20160422_130958](https://cloud.githubusercontent.com/assets/1638128/14758929/f2453402-08dd-11e6-8157-1c60afa7ab91.jpg)

**Comment by mintplugins** [comment]
@easydigitaldownloads/core-devs In order for Paypal Adaptive to use the same calculation function as commissions (which I think is important to make sure they don't every get out of sync), we need to do the commission calculation prior to the payment being completed. Currently the commission calculation is being completed after payment is complete through the "edd_update_payment_status" hook. Is there a similar hook that fires before payment is completed - and before sending it off to the payment gateway that gives access to the $cart_details?

**Comment by pippinsplugins** [comment]
@mintplugins Not following completely. Why can't we use the same one already?

**Comment by mintplugins** [comment]
Spoke with Pippin about this on slack but just leaving a note here as well:

The commission calculation/creation function doesn't actually run until after the payment gets marked as complete. However, we need the payment calculation BEFORE we head off to Paypal Adaptive Payments. 

The way I will go about resolving this is to write a separate helper function for the commission calculations which will be used by both the Paypal Adaptive function **(before payment)** and the Commissions function for creating the actual Commission posts after the payment comes back from Paypal Adaptive and is set to 'complete' **(after payment).**

**Comment by mintplugins** [comment]
New commissions calculation helper function, new Paypal Adaptive calculations, and Simple Shipping functions: https://github.com/easydigitaldownloads/EDD-Commissions/pull/130

**Comment by mintplugins** [comment]
Customer is testing https://github.com/easydigitaldownloads/EDD-Commissions/pull/130

https://secure.helpscout.net/conversation/167433619/26990/?folderId=180504

**Comment by pippinsplugins** [comment]
@mintplugins I see the error the customer reported. Ready for me to test further or wait?

**Comment by mintplugins** [comment]
This is ready for another round of testing

**Comment by pippinsplugins** [comment]
@mintplugins Are the breakdowns shown in your drawing above still accurate?

**Comment by mintplugins** [comment]
@pippinsplugins  No not totally. Cases 1, 2, and 3 should be the same. Case 4 will give the remaining percentage to the site owner. In that example it would/should give 98% to the site owner and 1% to each vendor.

**Comment by mintplugins** [comment]
Another customer testing here: 

https://secure.helpscout.net/conversation/249156666/41960/?folderId=180504

**Comment by pippinsplugins** [comment]
@mintplugins ping me when 4 is ready too.

**Comment by mintplugins** [comment]
@pippinsplugins This is ready. Essentially, for Case 4, if a user with the Store's email is set to get 1% in commissions for a product, it will get converted to 97% (with V1 and V2 getting 1% each as well). 

While the user with the store's email will still correctly earn/register a "Commission" of 1%, in Paypal Adaptive, the store's email will get 97%.

**Comment by mintplugins** [comment]
Here are some of the things I believe should be tested to ensure all is working. I have tested these myself but would like other people to test if possible. These need to be compeltely tested to make sure that Paypal actually splits up the payments correctly. Make sure that in chained mode the store is always the primary receiver and that it pays out vendors from there.

### Paypal Adaptive in Chained Mode:

**1) Store earns nothing:**

- Store Owner: 0%
- Vendor 1: 50%
- Vendor 2: 50%

**2) Store is also vendor:**

- Store Owner/Vendor: 50%
- Vendor 1: 50%

**3) Store earns leftovers:**
- Store Owner: 10%
- Vendor 1: 40%
- Vendor 2: 40%

### Paypal Adaptive in Parallel Mode:

**1) Store earns nothing:**
- Vendor 1: 50%
- Vendor 2: 50%

**2) Store is also vendor:**
- Store Owner/Vendor: 50%
- Vendor 1: 50% 

**3) Store earns leftovers:**
- Store Owner: 10%
- Vendor 1: 40%
- Vendor 2: 40%

**Comment by pippinsplugins** [comment]
Excellent. On my priority list for next week.

**Comment by pippinsplugins** [comment]
@mintplugins I don't think parallel payments are working. There's no error message but after going to PayPal it still shows only the one receiver (the main store):
![screen shot 2016-09-21 at 12 02 40 pm](https://cloud.githubusercontent.com/assets/1034109/18720841/564c0098-7ff3-11e6-9cc6-fc93b8db3b0c.png)
It's supposed to show each payment recipient.

Do you get the same behavior?

**Comment by pippinsplugins** [comment]
Wait, ignore me. I didn't have instant pay enabled :D

**Comment by pippinsplugins** [comment]
This all appears to work really well. All calculations check out for me!

Two main things I've found that still need updates:

1. We need to add some extensive help text to the settings and the new shipping options in the commissions metabox of the product edit screen. These options are complicated and there are quite a few possible configurations so lets get some good help text put in. The help text should also link to live documentation where we can put more advanced guides / walk throughs.

2. When instant pay is enabled, commission records are supposed to be immediately marked as `Paid` when they are created. This does not seem to be happening. All commissions are coming in as `Unpaid`.

**Comment by mintplugins** [comment]
1. I wrote a new doc for the Integration Guide between Commissions and Simple Shipping:
https://secure.helpscout.net/docs/5488f10de4b0dc8d3cacdf29/article/57e45200c697910d0784dcf9

I am working on the tool tips and help text in the settings. They should be ready tomorrow. 

I don't believe we need to mention Paypal Adaptive with this integration. While it integrates automatically now, there are no specific setting that correlate Simple Shipping and Paypal Adaptive.

2. I tested this again and my commissions are being marked as paid when using Paypal Adaptive Payments with Instant Pay enabled in the Commissions settings.

**Comment by mintplugins** [comment]
@pippinsplugins I have added tool-tips and improved some of the description language as well as provided a direct link to the (currently unpublished) integration doc on helpscout. I feel that linking directly to the doc is best as the amount of text needed to describe this properly will only clutter the metabox.

<img width="269" alt="screen shot 2016-09-23 at 1 22 46 pm" src="https://cloud.githubusercontent.com/assets/1638128/18795120/fbac03b4-8190-11e6-9621-e9f59b91e652.png">

**Comment by michaelbeil** [comment]
What about "To learn more about how this works, see [our doc](link)."

**Comment by mintplugins** [comment]
@michaelbeil I like it. Though my fear is that using the pronoun "our" can lose context when on a screen with other potential metaboxes we may not know about (other add-ons etc).

**Comment by michaelbeil** [comment]
`this doc` is fine.

**Comment by mintplugins** [comment]
@m 👍 https://github.com/easydigitaldownloads/EDD-Commissions/pull/130/commits/717be2ad5d5c4cb8ac9cc9940fb566bccb237468

**Comment by michaelbeil** [comment]
👌

**Comment by mintplugins** [comment]
Quick note that this will also solve this customer's problem:
https://secure.helpscout.net/conversation/192204548/32189/?folderId=749231

His issue was that if the "Store" is the single receiver in commissions (say a 99% commission) and the payment mode was "Chained" for Paypal Adaptive payments, the store was passed as both the first and second receiver which triggered an error in Paypal Adaptive. This update also fixes that problem by making sure each receiver sent to Paypal is unique (ie not sent as 2 different recievers in the list) and their amounts are calculated correctly. For reference, the specific error message there was "You must specify only one primary receiver and at least one secondary receiver".

**Comment by pippinsplugins** [comment]
@mintplugins could you put together the documentation page for this?

**Comment by mintplugins** [comment]
@pippinsplugins Already done :) 

https://github.com/easydigitaldownloads/EDD-Commissions/issues/125#issuecomment-249039634

Doc is ready for review here: https://secure.helpscout.net/docs/5488f10de4b0dc8d3cacdf29/article/57e45200c697910d0784dcf9

**Comment by pippinsplugins** [comment]
@mintplugins I've just run through a bunch of tests here and everything is working really nicely. I tested every variation I could think of and it all worked great! And unit tests pass :)

We'll need a sign off from @cklosowski and @SDavisMedia before we can merge this into `release/3.5` but looks like it should be good to go!

Nice work!

**Comment by mintplugins** [comment]
So good to hear! Thank you!

**Comment by SDavisMedia** [comment]
@mintplugins This is on my radar. Been in an open tab for a few days. I should be able to read up and test soon.

**Comment by mintplugins** [comment]
@SDavisMedia Just a little bump. If you end up with some spare cycles I'd love to get your eyes on this.

**Comment by ghost1227** [comment]
@pippinsplugins This gets a +1 from me! Works great! (can we get this merged sooner rather than later, it's gonna cause killer merge conflicts if we wait long)

**Comment by pippinsplugins** [comment]
Just waiting on a test from @SDavisMedia and a code review from @cklosowski

**Comment by ghost1227** [comment]
I'd like to get this tagged as a priority. It's going to wreak havoc with a ton of the other open issues when it comes to merge conflicts. Better to get it merged first if we can.

**Comment by SDavisMedia** [comment]
Yea I'm on it today.

**Comment by ghost1227** [comment]
Thanks sean!

**Comment by cklosowski** [comment]
Code review complete, some changes requested.

**Comment by mintplugins** [comment]
@cklosowski Thanks :) Changes committed!

**Comment by cklosowski** [comment]
If this is all working for you as expected, I'm good with the code changes from a code perspective.

**Comment by mintplugins** [comment]
@cklosowski Ran another test and it is all working for me here.

**Comment by mintplugins** [comment]
@SDavisMedia Sorry to be a pain but I just want to keep this on your radar if you get a spare moment.

**Comment by SDavisMedia** [comment]
Oh hi guys was I supposed to be testing something? :P 

-- 

Alright I tested everything I could think of. Here's what I found:

#### Commissions Only

All calculations look good with all settings. I tested percentage commissions, flat commissions, multiple users receiving commissions both percentage and flat ... also calculations based on Subtotal, total with taxes, and total without taxes. Everything checks out for me.

#### Commissions + Simple Shipping

Same as above. All checks out.

#### Commissions + PayPal Adaptive Payments + Simple Shipping

_Everything here looks great except for flat rate commissions. Now I'm pretty sure that's by design. But if not, flat rate commissions on the product settings will completely ignore shipping fee transfers to the users. Users will only receive their commissions regardless of settings when the commission settings are flat rate. That's the only problem I saw._

When testing with percentage commissions, everything works out perfectly with one or more users receiving commissions. I tested with and without discounts, with and without taxes, all global commission + shipping settings and all product commission + shipping settings. Everything looks good. Very impressive.

**Comment by mintplugins** [comment]
@SDavisMedia Awesome stuff. Thanks for making these tests happen. Glad to hear everything is working as expected. 

Regarding Flat Rate Commissions, that is by design. To avoid user confusion, what should probably happen is that the Commissions+Shipping settings only show if Shipping is enabled and the Commission Type is percentage. What we could do is load a new JS file only if Shipping and Commissions are both enabled which would show/hide the shipping setting accordingly. 

For now I have added a note to the docs letting users know that Flat Rate commissions will not be split up. 

Regarding the addition of new JS, my vote would be to add this if/when we find users getting confused by it. 

Again, awesome testing @SDavisMedia - thanks for making it happen!

**Comment by mintplugins** [comment]
@easydigitaldownloads/core-devs Let's do a vote:

🎉 = Add JS now to hide/show shipping options only when they make sense
👍 = To merge this as-is and add JS if/when customers come to us with confusion about flat rates.

**Comment by mintplugins** [comment]
JS Added. It will only show the Shipping Split options if all of the following are true:

1. Shipping is enabled
2. Commissions are enabled
3. Percentage Type is selected.

@SDavisMedia If you could do one more quick test of this that would be awesome. 

Commit:
https://github.com/easydigitaldownloads/EDD-Commissions/pull/130/commits/31e7aa3e777c5fb0004b3b1d31980990a21cee7e

**Comment by SDavisMedia** [comment]
@mintplugins Everything's looking good my way. Can't break it. :+1:

**Comment by mintplugins** [comment]
@easydigitaldownloads/core-devs Officially requesting merge of: 

https://github.com/easydigitaldownloads/EDD-Commissions/pull/130

**Comment by pippinsplugins** [comment]
Fantastic!

**Comment by mintplugins** [comment]
Merged! 🎉 🎉 🎉 🎉 🎉
